### ISSUE TITLE
Add JobSet SuccessPolicy 

### ISSUE THREAD

**Issue by ahg-g** [kubernetes-sigs/jobset#82]
Currently JobSet is marked as successful when all child jobs do.

We want to add success policy that allows customizing this behavior as originally discussed in the proposal doc https://bit.ly/k8s-jobset

A possible API may look like this:

```
type Operator string

// TerminationPolicyTargetAny applies to any job in the JobSet.
const OperatorAll Operator = "All"
const OperatorAny Operator = "Any"

type SuccessPolicy struct {
         // Operator determines either All or Any of the selected jobs should succeed to consider the JobSet successful
         Operator Operator

         // JobSelector selects the jobs that the policy will apply to. The selector only applies to the jobs created by the JobSet.
	JobSelector *metav1.LabelSelector         
}
```

**Comment by danielvegamyhre** [comment]
@ahg-g Rather than having the user configure a label selector to select which jobs to apply the success policy to, could we just have them specify which ReplicatedJobs to apply the policy to? 

I understand the label selector is more flexible, but for the most common cases of "succeed when all jobs succeed" or "succeed when leader succeeds," it seems it complicates things by requiring the user to know that we apply the replicatedJob name as a label, so that they can select `jobset.sigs.k8s.io/replicatedjob-name=leader` etc. so they'll need to dig through the docs to figure out how to do this.

I think a simplified API that is immediately obvious how to use might be more user friendly:

```
type SuccessPolicy struct {
         // Operator determines either All or Any of the selected jobs should succeed to consider the JobSet successful
         Operator Operator

         // Names of Replicated Jobs for the success policy to apply to (e.g. all jobs in the "worker" ReplicatedJob, or any job in the "leader" ReplicatedJob, etc.)
         // An empty list will apply the success policy to all ReplicatedJobs.
	ReplicatedJobNames []string 
}
```

Is there any specific use case the increased flexibility of the label selector is needed to support?

**Comment by ahg-g** [comment]
A list was the original API proposal as you know. However, now that we have a replicated list of jobs, I wanted to cover the case where the user wants to select a specific job or set of job replicas (e.g., job index 0).

As for ease of use, a label selector is the canonical API for object selection in k8s. Also, the user doesn't need to lookup the label that JobSet applies, they can use their own label too.

**Comment by danielvegamyhre** [comment]
I see, that makes sense. I've implemented this #124, let me know if you have any thoughts on my approach to the implementation.
